### PR TITLE
On demand summarize

### PR DESCRIPTION
**Pull request by arinwt**
This the final PR as part of #6382 (Part 3/3)
PR #6788 should be merged first.

Get return results from summarize attempts within SummaryGenerator, and then propagate those up through a new function summarizeOnDemand.

The RunningSummarizer class actually takes care of the work for generating the on demand summary. So it can only be done from a summarizer client, but we expose it all the way up to the parent client.

### Possible call methods
There are two possible ways to call this, pivoting on the clientDetails.type being "summarizer" or not.
####  Call on interactive parent client (type !== "summarizer")
1. `ContainerRuntime.summarizeOnDemand` on parent client
2. `SummaryManager.summarizeOnDemand` on parent client
    - fails if this client is not tracking a spawned summarizer client
3. `Summarizer.summarizeOnDemand` on summarizer client
    - fails if not started yet (via `run()` call) or if disposed/stopped
4. `RunningSummarizer.summarizeOnDemand` on summarizer client
    - returns a promise to await if a summary is already in progress

#### Call on non-interactive summarizer client (type === "summarizer")
1. `ContainerRuntime.summarizeOnDemand` on summarizer client
2. `Summarizer.summarizeOnDemand` on summarizer client
    - fails if not started yet (via `run()` call) or if disposed/stopped
3. `RunningSummarizer.summarizeOnDemand` on summarizer client
    - returns a promise to await if a summary is already in progress

So as you can see, it bypasses (2) if you have access directly to the summarizer container. This is intended for testing purposes.

### Result type
The `summarizeOnDemand` function synchronously returns an object with several promises that can be awaited as needed.
- If something goes wrong in any of the piping functions, then an error will be thrown with a brief message.
- If the innermost function detects that it is already generating a summary, it will return an object with an `alreadyRunning` property, which is a promise that resolves when the already running summary attempt completes.
- If the `alreadyRunning` property is undefined, then we attempt to generate an on-demand summary, and we get the object containing the following promises:
    - generateSummary: this promise resolves successfully as soon as the summary tree is generated, uploaded to storage, and the op is submitted. It may still return some information even if one of those steps fails. It will contain the summary tree, handle from storage, and client sequence number of submitted op.
    - broadcastSummaryOp: this promise resolves successfully as soon as the summarizer client observes its summarize op broadcasted from the server. It will contain the "summarize" op message.
    - summaryAckNack: this promise resolves when all the summarizing work is complete, indicated by either a summary ack or nack op. It will contain the "summaryAck" or "summaryNack" op message.

### Future to-dos
Probably should add more unit testing, specifically around the new SummaryGenerator.

Should further clean up the error processing within summarize flow, particularly within Summarizer.

Some other possible actions:
- When calling on interactive parent client: could spawn summarizer client on demand call, if not already running
- When calling on non-interactive summarizer client: could start running summarizer on demand call, if not already running

### CODE REVIEW

File: packages/runtime/container-runtime/src/runningSummarizer.ts [hunk 1]
```diff
@@ -279,14 +284,37 @@ export class RunningSummarizer implements IDisposable {
...
+        if (this.summarizingLock !== undefined) {
+            // The heuristics are blocking concurrent summarize attempts.
+            return { alreadyRunning: this.summarizingLock };
+        }
+        if (this.generator.isSummarizing()) {
+            // Another summary is currently being generated.
+            return { alreadyRunning: this.generator.waitSummarizing() };
+        }
+        const result = this.generator.summarize(`onDemand;${reason}` as `onDemand;${string}`, options);
+        result.summaryAckNack.finally(() => this.checkRerunHeuristics());
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/runningSummarizer.ts:308]: Why is this needed? Is it because while we are doing onDemandSummary, summarize was attempted via heuristics? Could you please also add a comment about this.

File: packages/runtime/container-runtime/src/summarizerTypes.ts [hunk 2]
```diff
@@ -132,6 +133,43 @@ export type GenerateSummaryResult =
...
+} | {
+    success: false;
+    data: T | undefined;
+    message: string;
+    error: any;
+};
+
+export interface ISummarizeResult {
+    /** Resolves when we generate, upload, and submit the summary */
+    readonly generateSummary: Promise<SummarizeResultPart<GenerateSummaryResult>>;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summarizerTypes.ts:158]: GenerateSummaryResult does not include the summaryTree that was generated. For testing, we will want access to the summaryTree right.

File: packages/runtime/container-runtime/src/summarizerTypes.ts [hunk 3]
```diff
@@ -132,6 +133,43 @@ export type GenerateSummaryResult =
...
+    /** Resolves when we see our summarize op broadcast; is sequence number of op */
+    readonly broadcastSummaryOp: Promise<SummarizeResultPart<IBroadcastSummaryResult>>;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summarizerTypes.ts:160]: nit: Maybe this should be called `broadcastSummaryResult` or just `broadcastSummary` as it returns more than the op.

File: packages/runtime/container-runtime/src/summarizerTypes.ts [hunk 4]
```diff
@@ -132,6 +133,43 @@ export type GenerateSummaryResult =
...
+    success: false;
+    data: T | undefined;
+    message: string;
+    error: any;
+};
+
+export interface ISummarizeResult {
+    /** Resolves when we generate, upload, and submit the summary */
+    readonly generateSummary: Promise<SummarizeResultPart<GenerateSummaryResult>>;
+    /** Resolves when we see our summarize op broadcast; is sequence number of op */
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summarizerTypes.ts:159]: Not sure what "is sequence number of op" means.

File: packages/runtime/container-runtime/src/summarizerTypes.ts [hunk 5]
```diff
@@ -132,6 +133,43 @@ export type GenerateSummaryResult =
...
+    readonly broadcastSummaryOp: Promise<SummarizeResultPart<IBroadcastSummaryResult>>;
+    /** True for ack; false for nack */
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summarizerTypes.ts:161]: What is true for ack and false for nack here?

File: packages/runtime/container-runtime/src/runningSummarizer.ts (hunk 1 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/runningSummarizer.ts:308]: Yes, that is correct reasoning. Basically rerunning heuristics afterwards. I'll add a comment.

File: packages/runtime/container-runtime/src/summarizerTypes.ts (hunk 2 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizerTypes.ts:158]: Oops! I thought the stats type had the tree too... good catch; will fix!

File: packages/runtime/container-runtime/src/summarizerTypes.ts (hunk 3 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizerTypes.ts:160]: So the name isn't about what it returns, but about which stage in the process it represents.
So the first stage is to generate the summary. The second stage is to wait for the summary op to be broadcast. And the third stage is to wait for the summary ack to be received.

That being said, I also am not a huge fan of the naming here.. I'll rethink the names.

File: packages/runtime/container-runtime/src/summarizerTypes.ts (hunk 4 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizerTypes.ts:159]: Outdated comment... Will remove. Thanks!

File: packages/runtime/container-runtime/src/summarizerTypes.ts (hunk 5 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizerTypes.ts:161]: Also outdated comment. Will remove.

File: packages/runtime/container-runtime/src/summarizerTypes.ts (hunk 2 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizerTypes.ts:158]: Added the summaryTree and gcData.

File: packages/runtime/container-runtime/src/summarizerTypes.ts (hunk 3 above)
Comment (arinwt) [reply packages/runtime/container-runtime/src/summarizerTypes.ts]: I changed it. Let me know what you think.

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/summarizerTypes.ts]: Looks much better.

**Comment by agarwal-navin** [comment]
We need tests to validate that on-demand summary is working as expected. That will also help us identify if we need to add / update things to make it easier to test. You could probably look at updating existing summarize / GC tests to use on demand summary.

**Comment by agarwal-navin** [comment]
I was thinking about the following scenario that we have discussed before -
For testing, most of the time we want to summarize after certain operations have happened. Basically, something like wait for a summary that includes sequence number X. Currently, the tests do that via `SummaryCollection`. They wait for a summary ack for a summary that contains X. I was wondering if it might be worth building that into the on demand summary - Basically, you can specify that you want a summary that contains X.
Do you think we should add that to this current API or build a layer on top of this for testing that can do the above? Or perhaps we can continue doing what the tests are currently doing?

**Comment by arinwt** [comment]
I was thinking about the seq number summarization thing, and I figure I can add it in a follow-up PR. I started tinkering with it already, and I think I should be able to add it to the options object.

**Comment by agarwal-navin** [comment]
Sounds good :-).

**Comment by agarwal-navin** [comment]
Is there a way to disable summarize heuristics? Tests would want to do that and just run summarize on demand. I believe that is what the work ask is too.

**Review by agarwal-navin — APPROVED** [review]
Looks good to me.

**Comment by arinwt** [comment]
Disable heuristics and queueing up a summary at a sequence number will be later PRs.
